core/include/edm/container.hpp

```diff
@@ -51,9 +84,163 @@ namespace traccc {
 	/// The item vector type
         using item_vector = jagged_vector_type< item_t >;
 	
+        /**
+         * @brief The type name of the element view which is returned by various
+         * methods in this class.
+         */
+        using element_view = container_element<
+            header_t,
+            typename item_vector::value_type
+        >;
+
+        /**
+         * @brief The size type of this container, which is the type by which
+         * its elements are indexed.
+         */
+        using size_type = typename header_vector::size_type;
         /// @}
 
+        /**
+         * We need to assert that the header vector and the outer layer of the
+         * jagged vector have the same size type, so they can be indexed using
+         * the same type.
+         */
+        static_assert(
+            std::is_same_v<
+                typename header_vector::size_type,
+                typename item_vector::size_type
+            >,
+            "Size type for container header and item vectors must be the same."
+        );
+
+        /**
+         * @brief Standard two-argument constructor.
+         *
+         * To enforce the invariant that both vectors must be the same size, we
+         * check this in the constructor. This is also checked in release
+         * builds.
+         */
+        container(
+            header_vector && hv,
+            item_vector && iv
+        )
+            : headers(hv)
+            , items(iv)
+        {
+            if (headers.size() != items.size()) {
+                throw std::logic_error("Header and item length not equal.");
+            }
+        }
+
+        /**
+         * @brief Argument-forwarding constructor.
+         *
+         * This constructor forwards its arguments to both internal vectors.
+         * Obviously, this requires the arguments to be quite general, but this
+         * constructor can be useful when initializing a container for a
+         * specific vecmem memory resource.
+         */
+        template<class... args>
+        container(
+            args&&... a
+        )
+            : headers(a...)
+            , items(a...)
+        {}
+
+        /**
+         * @brief Bounds-checking mutable element accessor.
+         */
+        element_view
+        at(
+            const size_type & i
```

Comment:
😕  Does the standard library use size type variables like this? I don't think we should be using constant references for these arguments. The compiler should have a much easier time with size variables passed simply by value.

Reply:
The standard library uses both const references and pass-by-value depending on the container. Since const references are more universal than pass by value, I figure that this would be a more general solution.

Reply:
Let's just follow what `std::vector` does. And that just passes size variables around by value.

https://en.cppreference.com/w/cpp/container/vector/at

We don't want to use anything other than integer type indices on these containers.

Reply:
Sure! If only we had C++20 and we could enforce a concept equal to the SequenceContainer named requirement... :cry: